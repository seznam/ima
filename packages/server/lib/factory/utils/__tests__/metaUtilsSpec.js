const { renderMeta } = require('../metaUtils');

describe('metaUtils', () => {
  describe('renderMeta', () => {
    it('should return empty string if the meta manager is not defined', () => {
      expect(renderMeta(undefined)).toBe('');
    });

    it('should generate meta tags while filtering out invalid values', () => {
      const meta = renderMeta({
        getTitle: jest.fn().mockReturnValue('Page title'),
        getMetaNamesIterator: jest.fn().mockReturnValue(
          new Map([
            ['description', { content: 'meta description' }],
            [
              'keywords',
              {
                content: 'ima core cli',
                'data-custom-attr': 'custom-attribute',
              },
            ],
          ]).entries()
        ),
        getMetaPropertiesIterator: jest.fn().mockReturnValue(
          new Map([
            ['og:description', { content: null }],
            [
              'og:keywords',
              {
                content: 'ima core cli',
                'data-custom-attr': 'custom-attribute',
                invalid: null,
              },
            ],
          ]).entries()
        ),
        getLinksIterator: jest
          .fn()
          .mockReturnValue(
            new Map([
              [
                'stylesheet',
                { href: '/media/examples/link-element-example.css' },
              ],
            ]).entries()
          ),
      });

      expect(meta).toMatchSnapshot();
      expect(meta.match(/data-ima-meta/g)).toHaveLength(5);
    });

    it('should return empty string for empty meta manager', () => {
      const meta = renderMeta({
        getTitle: jest.fn().mockReturnValue(''),
        getMetaNamesIterator: jest.fn().mockReturnValue(new Map([]).entries()),
        getMetaPropertiesIterator: jest
          .fn()
          .mockReturnValue(new Map([]).entries()),
        getLinksIterator: jest.fn().mockReturnValue(new Map([]).entries()),
      });

      expect(meta).toBe('<title></title>');
    });

    it('should generate meta tags while prevent XSS', () => {
      const meta = renderMeta({
        getTitle: jest.fn().mockReturnValue('<script>alert(1)</script>'),
        getMetaNamesIterator: jest.fn().mockReturnValue(
          new Map([
            ['description', { content: '"><script>alert(1)</script>' }],
            [
              'keywords',
              {
                content: '"><script>alert(1)</script>',
                'data-custom-attr': 'custom-attribute',
              },
            ],
            [
              '"><script>alert(1)</script>',
              {
                content: '"><script>alert(1)</script>',
                'data-custom-attr': 'custom-attribute',
              },
            ],
          ]).entries()
        ),
        getMetaPropertiesIterator: jest.fn().mockReturnValue(
          new Map([
            ['og:description', { content: null }],
            [
              'og:keywords',
              {
                content: '"><script>alert(1)</script>',
                'data-custom-attr': 'custom-attribute',
                invalid: null,
              },
            ],
          ]).entries()
        ),
        getLinksIterator: jest.fn().mockReturnValue(
          new Map([
            [
              'stylesheet',
              { href: '/media/examples/link-element-example.css' },
            ],
            ['stylesheet', { href: '"><script>alert(1)</script>' }],
          ]).entries()
        ),
      });

      expect(meta).toMatchSnapshot();
      expect(meta.match(/data-ima-meta/g)).toHaveLength(6);
    });
  });
});
