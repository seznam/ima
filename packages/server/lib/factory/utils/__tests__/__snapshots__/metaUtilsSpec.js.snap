// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metaUtils renderMeta should generate meta tags while filtering out invalid values 1`] = `"<title>Page title</title><link data-ima-meta rel="stylesheet" href="/media/examples/link-element-example.css" /><meta data-ima-meta name="description" content="meta description" /><meta data-ima-meta name="keywords" content="ima core cli" data-custom-attr="custom-attribute" /><meta data-ima-meta property="og:description" /><meta data-ima-meta property="og:keywords" content="ima core cli" data-custom-attr="custom-attribute" />"`;

exports[`metaUtils renderMeta should generate meta tags while prevent XSS 1`] = `"<title>&lt;script>alert(1)&lt;/script></title><link data-ima-meta rel="stylesheet" href="&quot;>&lt;script>alert(1)&lt;/script>" /><meta data-ima-meta name="description" content="&quot;>&lt;script>alert(1)&lt;/script>" /><meta data-ima-meta name="keywords" content="&quot;>&lt;script>alert(1)&lt;/script>" data-custom-attr="custom-attribute" /><meta data-ima-meta name="&quot;>&lt;script>alert(1)&lt;/script>" content="&quot;>&lt;script>alert(1)&lt;/script>" data-custom-attr="custom-attribute" /><meta data-ima-meta property="og:description" /><meta data-ima-meta property="og:keywords" content="&quot;>&lt;script>alert(1)&lt;/script>" data-custom-attr="custom-attribute" />"`;
